### PR TITLE
KAFKA-16190: Member should send full heartbeat when rejoining

### PR DESCRIPTION
**Pull request by phong260702**
When the consumer rejoins, heartbeat request builder make sure that all fields are sent in the heartbeat request.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: The current assignment should be empty when joining, so we don't technically need this. But we need to update `sentFields`! This should be fixed if you apply my suggetion from above

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: It's a bit weird, that in the JOINING state, we will update the data twice.

Maybe definiting a boolean `sendAllFields` and using it below would be clearer.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java
Comment (lucasbru) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:431]: The ticket asks for an extension of this test: 

> Note that the existing HeartbeatRequestManagerTest.testHeartbeatState misses this exact case given that it does explicitly change the subscription when it gets fenced.

**Review by lucasbru — CHANGES_REQUESTED** [review]
Thansk for the PR! I left some comments

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: nit: "Send"

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (phong260702) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: Thanks for the suggestion, I've changed it you take a look at it now.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java
Comment (phong260702) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:431]: I've add the test but not so sure if this is what it suppose to be can you check.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java
Comment (lucasbru) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java]: Hey @phong260702 . Sorry, I lost track of this PR. Feel free to ping me with my handle @lucasbru if that happens, I sometimes get lost in the GitHub noise.

I think the point of the comment was to have a case where use `transitionToFenced` but _not_ update the subscription. It seems you case does update the subscription.

Comment (phong260702) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java]: Oh I see. But isn't the test above with 'topic1' already done that?

Comment (lucasbru) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java]: Looks like it's updating the subscription

```
        String topic = "topic1";
        subscriptions.subscribe(Collections.singleton(topic), Optional.empty());
        membershipManager.onSubscriptionUpdated();
        membershipManager.transitionToFenced(); // And indirect way of moving to JOINING state
```

Comment (phong260702) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java]: Oh I see. I;m a little bit confuse about the comment. I'll fix it.

Comment (phong260702) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java]: @lucasbru Should we use `transitionToFenced` as it will move the state to JOINING instead of FENCED

Comment (lucasbru) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java]: yes. The ticket is about joining right? transitionTofenced is just a way to enter the rejoining state

Comment (phong260702) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java]: Got it. I've push a fix. Can you look at that.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: Please don't submit white-space only changes unless you are otherwise editing that fragment of the code.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java
Comment (lucasbru) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java]: The test is still subscribing to a new topic

Comment (phong260702) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java]: Forgot that mybad. Fixed that.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (lucasbru) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: This one is still open

Comment (phong260702) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: Fixed.

**Review by lucasbru — APPROVED** [review]
LGTM. Waiting for CI run before merging it

**Pull request merged**
